core/mcp.go

```diff
@@ -632,103 +761,147 @@ Each parameter corresponds to a named result with a specific purpose. Do not cal
-							return nil, fmt.Errorf("incompatible types: %s must be %s, got %s", name, expectedType, actualType)
-						}
+				expected := output.expectedType
+				expectedType := expected.TypeName()
+				if expectedType == "String" {
+					output.Value = dagql.String(argStr)
+				} else {
+					bnd, ok := m.env.objsByID[argStr]
+					if !ok {
+						return nil, fmt.Errorf("object not found for argument %s: %s", name, argStr)
+					}
+					obj := bnd.Value
+					actualType := obj.Type().Name()
+					if expectedType != actualType {
+						return nil, fmt.Errorf("incompatible types: %s must be %s, got %s", name, expectedType, actualType)
 					}
 					output.Value = obj
-				} else {
-					return nil, fmt.Errorf("undefined output: %q", name)
 				}
 			}
 			m.returned = true
 			return "ok", nil
 		},
-	}
+	}, true
 }
 
-func (m *MCP) Builtins(srv *dagql.Server) ([]LLMTool, error) {
-	builtins := []LLMTool{}
-
-	if len(m.env.outputsByName) > 0 {
-		builtins = append(builtins, m.returnBuiltin())
+func (m *MCP) Builtins(srv *dagql.Server, tools []LLMTool) ([]LLMTool, error) {
+	builtins := []LLMTool{
+		{
+			Name:        "think",
+			Description: `A tool for thinking through problems, brainstorming ideas, or planning without executing any actions`,
+			Schema: map[string]any{
+				"type": "object",
+				"properties": map[string]any{
+					"thought": map[string]any{
+						"type":        "string",
+						"description": "Your thoughts.",
+					},
+				},
+				"required": []string{"thought"},
+			},
+			Call: ToolFunc(func(ctx context.Context, args struct {
+				Thought string
+			}) (_ any, rerr error) {
+				ctx, span := Tracer(ctx).Start(ctx, "think",
+					telemetry.Reveal(),
+					trace.WithAttributes(
+						attribute.String(telemetry.UIMessageAttr, "thought"),
+						attribute.String(telemetry.UIActorEmojiAttr, "🤖"),
+					),
+				)
+				defer telemetry.End(span, func() error { return rerr })
+				stdio := telemetry.SpanStdio(ctx, InstrumentationLibrary,
+					log.String(telemetry.ContentTypeAttr, "text/markdown"))
+				defer stdio.Close()
+				fmt.Fprint(stdio.Stdout, args.Thought)
+				return "Finished thinking.", nil
+			}),
+		},
 	}
 
-	for _, typeName := range m.env.Types() {
-		tools, err := m.tools(srv, typeName)
-		if err != nil {
-			return nil, fmt.Errorf("tools for %q: %w", typeName, err)
-		}
+	if len(tools) > 0 {
 		builtins = append(builtins, LLMTool{
-			Name:    "select" + typeName,
-			Returns: typeName,
+			Name: "selectTools",
 			Description: (func() string {
-				desc := fmt.Sprintf("Select a %s by its ID.", typeName)
-				desc += "\n\nProvides the following tools:\n"
+				desc := `Select tools for interacting with the available objects.`
+				desc += "\n\nAvailable tools:"
 				for _, tool := range tools {
-					desc += fmt.Sprintf("\n- %s", tool.Name)
+					if m.selectedTools[tool.Name] {
+						// already have it
+						continue
+					}
+					desc += "\n- " + tool.Name + " (returns " + tool.Returns + ")"
+				}
```

Comment:
so with env(privileged:true) what tools are initially available in the tool description? literally all the core API, module and it's dependencies?

(i read some other code that made me think that this was dependent on input objects)

Reply:
Should be whatever's defined on `Query`

(pardon the terse replies, heading out)